include/envoy/network/filter.h

```diff
@@ -81,8 +81,9 @@ class ReadFilter {
   /**
    * Called when data is read on the connection.
    * @param data supplies the read data which may be modified.
-   * @param end_stream supplies whether this is the last byte on the connection. This will only
-   *        be set if the connection has half-close semantics enabled.
+   * @param end_stream supplies whether this is the last byte on the connection. It is possible that
```

Comment:
I dislike the interface that's being created here.  It's confusing to reason about, because we're saying that end_stream may or may not be delivered, depending on how the bytes are broken into packets. 

I think we should guarantee delivery of end_stream, so that it's simple to reason about the interface.

Reply:
I agree, but if we don't do this I really have no idea how we are going to "fix" the issue because of we supply an empty onData with end_stream as true, the original issue will still repro.

I still really think we should just not "fix" this at all...

Reply:
Maybe I should make a separate PR for `onRead` interface change since the discussion is separate.

@ggreenway not only depending on how the bytes are broken into packets, connection may be closed for several reasons. We cannot guarantee delivery of end_stream, instead, Envoy makes it best effort to set end_stream when `onData` is called, even if half-close is not enabled. I can revert source/common/network/connection_impl.cc line 439 but it still does not guarantee anything.

@mattklein123 I partly agree that for the original issue the upstream is responsible. While Envoy can make the best effort to let read filter know when it read end_stream (ReadFilter change), and HTTP conn pool can use that, no?

Reply:
@lizan I think we should always deliver end_stream if we can, ie if the connection is shutdown cleanly.

I think we revert line 439, and in the http handling, if you get onData(empty, true) you can ignore it.

Reply:
> @mattklein123 I partly agree that for the original issue the upstream is responsible. While Envoy can make the best effort to let read filter know when it read end_stream (ReadFilter change), and HTTP conn pool can use that, no?

Yes, but when we are spending this much time talking about this change, and the change is becoming increasingly risky, it seems not worth it to me. 

> @lizan I think we should always deliver end_stream if we can, ie if the connection is shutdown cleanly. I think we revert line 439, and in the http handling, if you get onData(empty, true) you can ignore it.

Yes I think this should work. It will "fix" the problem when we get FIN inline with the final data, but still always deliver an end_stream with zero data before closing the connection in the half close case. Though I really do worry that there are going to be other filters that won't correctly handle this.

One other option would be to add some type of function to connection such as `sawRemoteClose()` (similar to what we have on codec client), and allow codecClient to call this. Then we can keep the half close / end_stream logic as it currently is.

Reply:
> @lizan I think we should always deliver end_stream if we can, ie if the connection is shutdown cleanly. I think we revert line 439, and in the http handling, if you get onData(empty, true) you can ignore it.

Please see #2841, OTOH I think it might be better not deliver onData(0-length data, true) even half-close is enabled, and let filters rely on onEvent(RemoteClose) for that case. Since they are same and deliver onData doesn't let us win the race since no data is there.

> One other option would be to add some type of function to connection such as sawRemoteClose() (similar to what we have on codec client), and allow codecClient to call this. Then we can keep the half close / end_stream logic as it currently is.

I thought this once though I don't want to add another place in addition to `onData` and `onEvent` handle remote close. It should be simpler.